snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -584,18 +586,75 @@ def update_build(self):
             # file_utils.copy instead of link_or_copy, as the build process
             # may modify these files
             source = sources.Local(
-                self.plugin.sourcedir,
-                self.plugin.build_basedir,
-                copy_function=file_utils.copy,
+                self.part_source_dir, self.part_build_dir, copy_function=file_utils.copy
             )
             if not source.check(
-                states.get_step_state_file(self.plugin.statedir, steps.BUILD)
+                states.get_step_state_file(self.part_state_dir, steps.BUILD)
             ):
                 return
             source.update()
 
         self._do_build(update=True)
 
+    def _generate_env(self) -> str:
+        """
+        Generates an environment suitable to run during a step.
+
+        :returns: io.StringIO at the beginning of the stream.
+        """
+        if isinstance(self.plugin, plugins.v1.PluginV1):
+            raise RuntimeError("PluginV1 not supported.")
+
+        # TODO add Snapcraft's environment.
+
+        # Plugin's say.
+        plugin_build_environment = self.plugin.get_build_environment()
+
+        # Part's (user) say.
+        part_build_environment = self._part_properties["build-environment"]
+
+        # Create the script.
+        run_environment = io.StringIO()
+        print("#!/bin/sh", file=run_environment)
+
+        print("# Environment", file=run_environment)
+        print("## Plugin Environment", file=run_environment)
+        for env in plugin_build_environment:
+            for k, v in env.items():
+                print(f'export {k}="{v}"', file=run_environment)
+        print("## Part Environment", file=run_environment)
+        for env in part_build_environment:
+            for k, v in env.items():
+                print(f'export {k}="{v}"', file=run_environment)
+
+        run_environment.seek(0)
```

Comment:
you can use run_environment.getvalue() instead of `seek`/`read`.  i'd suggest using `with` to clean to make sure it is closed...

perhaps:
```
with io.StringIO() as run_environment:
  ..
  return run_environment.getvalue()
```

Reply:
good call